src/rust/demikernel/libos/mod.rs

```diff
@@ -201,6 +201,41 @@ impl LibOS {
         }
     }
 
+    /// Waits for an I/O operation to complete or a timeout to expire.
+    pub fn timedwait_any(
+        &mut self,
+        qts: &[QToken],
+        abstime: Option<SystemTime>,
+    ) -> Result<(usize, demi_qresult_t), Fail> {
+        trace!("timedwait_any() qts={:?}, timeout={:?}", qts, abstime);
+
+        loop {
+            // Poll first, so as to give pending operations a chance to complete.
+            self.poll();
+
+            // Search for any operation that has completed.
+            for (i, &qt) in qts.iter().enumerate() {
+                // Retrieve associated schedule handle.
+                // TODO: move this out of the loop.
+                let mut handle: SchedulerHandle = self.schedule(qt)?;
+
+                // Found one, so extract the result and return.
+                if handle.has_completed() {
+                    return Ok((i, self.pack_result(handle, qt)?));
+                }
+
+                // Return this operation to the scheduling queue by removing the associated key
+                // (which would otherwise cause the operation to be freed).
+                handle.take_key();
+            }
+
+            // Check if timeout has expired.
+            if abstime.is_none() || SystemTime::now() >= abstime.unwrap() {
+                return Err(Fail::new(libc::ETIMEDOUT, "timer expired"));
+            }
+        }
+    }
+
```

Comment:
Here too, `timedwait_any()` and `wait_any()` share almost all the same code.  `wait_any()` should either be implemented in terms of `timedwait_any`, or the latter should just be renamed to the former and have a single call.

Reply:
Thanks for noting this.

I had thought about unifying these calls as well. However, I follow a software development approach where I prefer to work with incremental changes in the code base. I believe that this is important specially because some design choices will become clearer with more iterations.

Sme thoughts we still should consolidate:
- What special value should it be, so it matches a natural programming pattern for Demikernel applications?
- Iif we decide to replace `timespec` for an integer that would give us the timeout value in nanoseconds similarly to `nanosleep()`? Then, what would be that speciall value

In any case I agree that this should be reconsidered in the future. I have created an issue to keep track of this on the `wait()` and `timedwait()` system calls. https://github.com/demikernel/demikernel/issues/316

Once `timedwait_any()` is merged in the source tree I create issue for the `timedwait_any()` and `wait_any()` calls as well.